src/server/pkg/storage/chunk/reader.go

```diff
@@ -0,0 +1,89 @@
+package chunk
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"io"
+	"path"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/grpcutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
+)
+
+// Reader reads a set of DataRefs from chunk storage.
+type Reader struct {
+	ctx      context.Context
+	objC     obj.Client
+	prefix   string
+	dataRefs []*DataRef
+	curr     *DataRef
+	buf      *bytes.Buffer
+	chunk    []byte
+	r        io.Reader
+}
+
+func newReader(ctx context.Context, objC obj.Client, prefix string, dataRefs []*DataRef) *Reader {
+	return &Reader{
+		ctx:      ctx,
+		objC:     objC,
+		prefix:   prefix,
+		dataRefs: dataRefs,
+		buf:      &bytes.Buffer{},
+		r:        bytes.NewReader([]byte{}),
+	}
+}
+
+// Read reads from the byte stream produced by the set of DataRefs.
+func (r *Reader) Read(data []byte) (int, error) {
+	var totalRead int
+	for len(data) > 0 {
+		n, err := r.r.Read(data)
+		data = data[n:]
+		totalRead += n
+		if err != nil {
+			// If all DataRefs have been read, then io.EOF.
+			if len(r.dataRefs) == 0 {
+				return totalRead, io.EOF
+			}
+			// Get next chunk if necessary.
+			if r.curr == nil || r.curr.Hash != r.dataRefs[0].Hash {
+				if err := r.nextChunk(r.dataRefs[0].Hash); err != nil {
+					return totalRead, err
+				}
+			}
+			r.curr = r.dataRefs[0]
+			r.dataRefs = r.dataRefs[1:]
+			r.r = bytes.NewReader(r.chunk[r.curr.Offset : r.curr.Offset+r.curr.Size])
+		}
+	}
+	return totalRead, nil
+
+}
+
+func (r *Reader) nextChunk(hash string) error {
+	objR, err := r.objC.Reader(r.ctx, path.Join(r.prefix, hash), 0, 0)
+	if err != nil {
+		return err
+	}
+	defer objR.Close()
+	gzipR, err := gzip.NewReader(objR)
+	if err != nil {
+		return err
+	}
+	defer gzipR.Close()
+	r.buf.Reset()
+	buf := grpcutil.GetBuffer()
+	defer grpcutil.PutBuffer(buf)
+	if _, err := io.CopyBuffer(r.buf, gzipR, buf); err != nil {
+		return err
+	}
+	r.chunk = r.buf.Bytes()
+	return nil
+}
+
+// Close closes the reader.
+// Currently a no-op, but will be used when streaming is implemented.
+func (r *Reader) Close() error {
+	return nil
+}
```

Comment:
The abstractions in this file are really nice. I have a few nits about the implementation though.

It feels kinda weird that `curr` is a `*DataRef` that continually gets popped off the `dataRefs` rather than just an index into it. Not a particularly strong feeling, I wouldn't say it's "wrong" as is, just not what I was immediately expecting.

The process for reading the data seems needlessly complication. You construct a reader for the object store, then copy all of its data into a `buf`, then extract the `[]byte` into `chunk` and finally pass `chunk` to `bytes.NewReader` (with `offset` applied). Can't this just be as simple as reading the data into `buf` and then calling `Seek` to apply the offset and using `buf` as our `Reader`? Also I think it would be good if this reading process happened entirely in `nextChunk` and if that function was renamed to `readChunk`, having `next` in the name suggests to me that it's responsible for advancing which chunk we're reading from, which doesn't really seem to be what it does. Similarly some parts of the reading process (such as applying the offset are in `Read`, they should probably be moved into `readChunk`.

Reply:
Using an index rather than a *DataRef is fine, but it is a bit more readable to just store a *DataRef than to constantly index into a slice. Also, it seemed to make sense to me to just pull data refs off the list until there are none left.

I don't think I fully understand your suggestion for simplifying this because I think you might be overlooking communicating the io.EOF (which wrapping with a bytes reader provides). If I don't wrap the data associated with a DataRef with a bytes reader, then I have to keep track of the bytes read. Using the bytes reader means I just have to setup the ranges and not think about the next one until that reader returns io.EOF. My full thoughts on this is that the bytes buffer is used as a general writer and memory pool, while the chunk / bytes readers are used for managing the reading of ranges within a chunk. I guess I do not really need to store the byte slice that corresponds to the chunk and instead just pull it from the buf when creating a new reader, so that might be cleaner.

Reply:
I might well be confused here, but won't `buf` return an `io.EOF` too once you read off the end of it?

Reply:
`buf` stores the full chunk, `r` wraps a range within a chunk that corresponds to a data ref (a single data ref is a range within a single chunk). If for example they have a set of data refs that all reference one chunk, but there are spaces between the referenced data, then we don't want them reading into those between spaces. There will definitely need to be more documentation on this, but I have been holding off because I imagine there is some more iteration left.

Reply:
Yeah, I don't think the things I'm suggesting with the `DataRef` are going to make things much more readable, so let's forget about those.

For the `Reader` stuff, what if you just wrapped the `bytes.NewReader` in a `LimitedReader` to implement the `EOF` where you need it? I think that might be a little bit cleaner, not enough that I think it's crucial though, so up to you.

Reply:
I don't think using a `LimitedReader` would make this cleaner because you would still need to skip the underlying reader ahead to setup the next read if you were to use the same reader. Just mapping the refs to byte reader ranges seems like the cleanest solution to me.

Reply:
I don't think you would need to skip it ahead, because the `LimitedReader` is just passing reads to the underlying `Reader`, so I think it would be right at the point you wanted it once the `LimitedReader` returned an `EOF`.